### PR TITLE
Fix #1490: JVM style module initialization under semantic flag

### LINKED ISSUE #1490: JVM style module initialization under semantic flag
Given this minimised ADT:
```scala
sealed trait A
object A {
  private def blankSym = ""

  sealed abstract class B extends A
  case object X extends B

  sealed abstract class C(symbol: String) extends A
  case object Y extends C(blankSym)

  val cs = Vector[C](Y)
}
```

At runtime, both `A.cs.head == A.Y` and `A.cs.head eq A.Y` will fail.

Guess how to make it work? Change
```scala
case object Y extends C(blankSym)
```
to
```scala
case object Y extends C("")
```

### PR DESCRIPTION
**Pull request by gzm0**

### CODE REVIEW

**Comment by gzm0** [comment]
Review by @sjrd.

**Comment by gzm0** [comment]
Maybe we should call the flag "moduleInits"?

File: test-suite/src/test/scala/org/scalajs/testsuite/compiler/ObjectInitTest.scala [hunk 1]
```diff
@@ -0,0 +1,37 @@
...
+import org.scalajs.jasminetest.JasmineTest
+import scala.scalajs.js
+
+object ObjectInitTest extends JasmineTest {
+
+  describe("Scala Modules") {
+    when("compliant-moduleinit").
+    it("Should only execute module initializers once") {
+      A.Y eq A.cs.head
+      expect(Counter.c).toBe(1)
```
Comment (sjrd) [inline test-suite/src/test/scala/org/scalajs/testsuite/compiler/ObjectInitTest.scala]: This should also test that the right one is actually a module, and the other one is actually `null`.

File: tools/shared/src/main/scala/org/scalajs/core/tools/javascript/ScalaJSClassEmitter.scala [hunk 2]
```diff
@@ -487,12 +487,41 @@ final class ScalaJSClassEmitter(semantics: Semantics, outputMode: OutputMode,
...
+            ),
+            js.Skip())
+        case CheckedBehavior.Fatal =>
+          js.If(moduleInstanceVar === js.Undefined(), {
+            js.Block(
+                moduleInstanceVar := js.Null(),
+                assignModule
+            )
+          }, js.If(moduleInstanceVar === js.Null(), {
+            // throw new UndefinedBehviorError(
```
Comment (sjrd) [inline tools/shared/src/main/scala/org/scalajs/core/tools/javascript/ScalaJSClassEmitter.scala]: typo

File: tools/shared/src/main/scala/org/scalajs/core/tools/javascript/ScalaJSClassEmitter.scala [hunk 3]
```diff
@@ -487,12 +487,41 @@ final class ScalaJSClassEmitter(semantics: Semantics, outputMode: OutputMode,
...
+            //     "Initializer of $className called before completion")
```
Comment (sjrd) [inline tools/shared/src/main/scala/org/scalajs/core/tools/javascript/ScalaJSClassEmitter.scala]: ... of its super constructor

File: tools/shared/src/main/scala/org/scalajs/core/tools/javascript/ScalaJSClassEmitter.scala [hunk 4]
```diff
@@ -487,12 +487,41 @@ final class ScalaJSClassEmitter(semantics: Semantics, outputMode: OutputMode,
...
+            val decodedName = Definitions.decodeClassName(className).stripSuffix("$")
+            val msg = s"Initializer of $decodedName called twice"
```
Comment (sjrd) [inline tools/shared/src/main/scala/org/scalajs/core/tools/javascript/ScalaJSClassEmitter.scala]: The text does not match the one in the comment. I prefer the one in the comment.

File: tools/shared/src/main/scala/org/scalajs/core/tools/sem/Semantics.scala [hunk 5]
```diff
@@ -16,7 +16,9 @@ import org.scalajs.core.tools.optimizer.LinkedClass
 final class Semantics private (
     val asInstanceOfs: CheckedBehavior,
     val strictFloats: Boolean,
-    val runtimeClassName: Semantics.RuntimeClassNameFunction) {
+    val runtimeClassName: Semantics.RuntimeClassNameFunction,
+    val moduleInit: CheckedBehavior
```
Comment (sjrd) [inline tools/shared/src/main/scala/org/scalajs/core/tools/sem/Semantics.scala]: I'd like to keep the CheckedBehaviors together, before `strictFloats`. The rest of the class should also follow that ordering.

File: ci/matrix.xml [hunk 6]
```diff
@@ -52,14 +52,22 @@
         'set scalaJSStage in Global := FastOptStage' \
         ++$scala testSuite/test \
         testSuite/clean &&
-    sbt 'set scalaJSSemantics in testSuite ~= { _.withAsInstanceOfs(org.scalajs.core.tools.sem.CheckedBehavior.Compliant).withStrictFloats(true) }' \
+    sbt 'set scalaJSSemantics in testSuite ~= { \
```
Comment (sjrd) [inline ci/matrix.xml]: You cannot use a `\` within a string, I believe.

**Comment by sjrd** [comment]
That's all.

**Comment by gzm0** [comment]
Updated

File: test-suite/src/test/scala/org/scalajs/testsuite/compiler/ObjectInitTest.scala [hunk 7]
```diff
@@ -0,0 +1,40 @@
...
+object ObjectInitTest extends JasmineTest {
+
+  describe("Scala Modules") {
+    when("compliant-moduleinit").
+    it("Should only execute module initializers once") {
+      val x = A.Y
+      val y = A.cs.head
+      expect(x eq null).toBeTruthy
+      expect(y eq A.Y).toBeTruthy
+      expect(Counter.c).toBe(1)
```
Comment (sjrd) [inline test-suite/src/test/scala/org/scalajs/testsuite/compiler/ObjectInitTest.scala]: Also test `y ne null`. The existing tests would succeed if everything returned `null` all the time.

File: project/Build.scala [hunk 8]
```diff
@@ -874,6 +874,11 @@ object Build extends sbt.Build {
         ) ++ (
             if (sems.strictFloats) Seq(Tests.Argument("-tstrict-floats"))
             else Seq()
+        ) ++ (
+            if (sems.moduleInit == CheckedBehavior.Compliant)
+              Seq(Tests.Argument("-tcompliant-moduleinit"))
+            else
+              Seq()
```
Comment (sjrd) [inline project/Build.scala]: To be consistent, these settings should also come between the `sems.asInstanceOfs` and `sems.strictFloats`.

File: tools/shared/src/main/scala/org/scalajs/core/tools/sem/Semantics.scala [hunk 9]
```diff
@@ -44,6 +49,7 @@ final class Semantics private (
     import scala.util.hashing.MurmurHash3._
     var acc = HashSeed
     acc = mix(acc, asInstanceOfs.hashCode)
+    acc = mix(acc, moduleInit.hashCode)
     acc = mixLast(acc, strictFloats.##)
     finalizeHash(acc, 1)
```
Comment (sjrd) [inline tools/shared/src/main/scala/org/scalajs/core/tools/sem/Semantics.scala]: 1 should 3 (apparently we already failed to turn it to 2 when we added `strictFloats`).

**Comment by sjrd** [comment]
That's all.

**Comment by gzm0** [comment]
Updated

**Comment by sjrd** [comment]
LGTM

**Comment by gzm0** [comment]
Updated with binary incompatibility filter.

**Comment by sjrd** [comment]
```
[error] Error during tests:
[error] 	org.scalajs.testsuite.compiler.ObjectInitTest
[error] (testSuite/test:test) sbt.TestsFailedException: Tests unsuccessful
```

**Comment by gzm0** [comment]
I understood the test the wrong way around. Luckily I implement how it makes sense to me and not test driven... :P I verified the current test against Scala JVM (2.11.4).

**Comment by sjrd** [comment]
LGTM
